components/brave_news/browser/feed_v2_builder.cc

```diff
@@ -1122,79 +1125,138 @@ void FeedV2Builder::NotifyUpdateCompleted() {
   std::tie(hash_, subscribed_count_) =
       GetFeedHashAndSubscribedCount(channels, publishers, feed_etags_);
 
-  // Fire all the pending callbacks.
-  for (auto& callback : current_update_->callbacks) {
-    std::move(callback).Run();
-  }
-
   // If we have a |next_update_| then request an update with that data.
+  auto callbacks = std::move(current_update_->callbacks);
   current_update_ = std::move(next_update_);
   next_update_ = std::nullopt;
 
-  // Notify listeners of updated hash.
-  for (const auto& listener : listeners_) {
-    listener->OnUpdateAvailable(hash_);
-  }
+  // Fire all the pending callbacks. Wait for them all to complete before we
+  // fire off the next update.
+  auto on_notify_completed = base::BarrierClosure(
+      callbacks.size(),
+      base::BindOnce(
+          [](base::WeakPtr<FeedV2Builder> builder) {
+            if (!builder) {
+              return;
+            }
 
-  if (current_update_) {
-    UpdateData(current_update_->settings);
+            // Notify listeners of updated hash.
+            for (const auto& listener : builder->listeners_) {
+              listener->OnUpdateAvailable(builder->hash_);
+            }
+
+            if (builder->current_update_) {
+              builder->UpdateData(
+                  builder->current_update_->settings,
+                  base::BindOnce([](base::OnceClosure on_complete) {
+                    std::move(on_complete).Run();
+                  }));
+            }
+          },
+          weak_ptr_factory_.GetWeakPtr()));
+
+  for (auto& callback : callbacks) {
+    std::move(callback).Run(on_notify_completed);
   }
 }
 
-void FeedV2Builder::GenerateFeed(
-    UpdateSettings settings,
-    mojom::FeedV2TypePtr type,
-    base::OnceCallback<mojom::FeedV2Ptr()> build_feed,
-    BuildFeedCallback callback) {
+void FeedV2Builder::GenerateFeed(UpdateSettings settings,
+                                 mojom::FeedV2TypePtr type,
+                                 FeedGenerator generate,
```

Comment:
optional: It'd be good to be consistent about naming

```suggestion
                                 FeedGenerator generator,
```